### PR TITLE
Improve query counts for CourseOffering.assignable_course_offerings

### PR DESCRIPTION
**Pull request by bethanyaconnor**
This is being done as part of the work to gather the quick assign course offerings. In general, our call to get the course offerings for a user is slow and makes a lot of queries to the database. This PR follows the same approach as https://github.com/code-dot-org/code-dot-org/pull/45432. The `course_offerings` table is currently seeded during the deploy and does not get updated on production, so we can use a class variable to store all instances. On environments where we do update this table (eg levelbuilder), we need to continue querying the database each time

I wasn't sure of a good way of testing performance improvements other than query count, so I used query count a proxy. I used a query counter gem locally for both `valid_course_offerings` -- the API used on the homepage currently -- and `quick_assign_course_offerings` -- the API used on the new section setup page. These are the number of uncached queries after querying the endpoint a few times.

|    | staging | this PR |
| - |---------|-------|
| valid_course_offerings | 1478 | 71 |
| quick_assign_course_offerings | 1340 | 46 |

### CODE REVIEW

File: dashboard/test/controllers/api/v1/sections_controller_test.rb [hunk 1]
```diff
@@ -1232,6 +1232,15 @@ def returned_section
     assert_nil @section.code_review_expires_at
   end
 
+  test 'num queries for valid_course_offerings for a teacher with no extra permissions' do
+    sign_in @teacher
+    Unit.stubs(:should_cache?).returns(true)
+
+    assert_cached_queries(73) do
```
Comment (bethanyaconnor) [inline dashboard/test/controllers/api/v1/sections_controller_test.rb:1239]: This was previously 126 so there was an improvement

File: dashboard/test/models/course_offering_test.rb [hunk 2]
```diff
@@ -437,6 +437,20 @@ class CourseOfferingTest < ActiveSupport::TestCase
     assert_equal facilitator_to_teacher_course_versions[@unit_facilitator_to_teacher.course_version.id][:units].keys, [@unit_facilitator_to_teacher.id]
   end
 
+  test 'query count for assignable_course_offerings' do
+    Unit.stubs(:should_cache?).returns true
+
+    assert_cached_queries(0) do
```
Comment (bethanyaconnor) [inline dashboard/test/models/course_offering_test.rb:443]: Worth mentioning that this one used to make 135 queries and the facilitator one made 136 queries.

File: dashboard/app/models/course_offering.rb [hunk 3]
```diff
@@ -104,8 +104,17 @@ def any_version_is_assignable_editor_experiment?(user)
     course_versions.any? {|cv| cv.content_root.is_a?(Unit) && cv.has_editor_experiment?(user)}
   end
 
+  def self.all_course_offerings
+    if should_cache?
+      @@course_offerings ||= CourseOffering.all.includes(course_versions: :content_root)
+      @@course_offerings
```
Comment (rshipp) [inline dashboard/app/models/course_offering.rb]: nit: i don't think this line is needed, since the one above should return the same thing?

**Review by rshipp — APPROVED** [review]
wow. that's a lot less queries

**Pull request merged**